src/sentry/integrations/github/blame.py

```diff
@@ -0,0 +1,213 @@
+        except ValueError:
+            continue
+
+        repo_queries += _make_repo_query(repo_name, repo_owner, ref_queries, repo_index)
+
+    return f"""query {{{repo_queries}\n}}"""
+
+
+def extract_commits_from_blame_response(
+    response: MappingApiResponse,
+    files: Sequence[SourceLineInfo],
+    file_path_mapping: FilePathMapping,
+    extra: Optional[Dict[str, str]] = None,
+) -> Sequence[FileBlameInfo]:
+    """
+    Using the file path mapping that generated the initial GraphQL query,
+    this function extracts all commits from the response and maps each one
+    back to the correct file.
+    """
+    file_blames: List[FileBlameInfo] = []
+    for repo_index, (full_repo_name, ref_mapping) in enumerate(file_path_mapping.items()):
+        repo_mapping: Optional[Dict[str, Any]] = response.get("data", {}).get(
+            f"repository{repo_index}"
+        )
+        if not repo_mapping:
+            logger.info(f"Repository {full_repo_name} does not exist in GitHub.", extra=extra)
+            continue
+        for ref_index, (ref_name, file_paths) in enumerate(ref_mapping.items()):
+            ref: Optional[Dict[str, Any]] = repo_mapping.get(f"ref{ref_index}")
+            if not ref:
+                logger.info(
+                    f"Branch {ref_name} for repository {full_repo_name} does not exist in GitHub.",
+                    extra=extra,
+                )
+                continue
+            for file_path_index, file_path in enumerate(file_paths):
+                blame: Optional[Dict[str, Any]] = ref.get("target", {}).get(
+                    f"blame{file_path_index}"
+                )
+                if not blame:
+                    logger.info(
+                        f"File {file_path} for branch {ref_name} for repository {full_repo_name} does not exist in GitHub.",
+                        extra=extra,
+                    )
+                    continue
+                matching_file, blame_info = _get_matching_file_and_blame(
+                    files=files,
+                    blame_ranges=blame.get("ranges", []),
+                    path=file_path,
+                    repo_name=full_repo_name,
+                    ref=ref_name,
+                )
+                if not blame_info:
+                    logger.info(
+                        f"No matching commit found for line {matching_file.lineno} in file {matching_file.path} in branch {matching_file.ref} in repository {full_repo_name} in GitHub.",
+                        extra=extra,
+                    )
+                    continue
+                file_blames.append(blame_info)
+    return file_blames
+
+
+def _get_matching_file_and_blame(
+    files: Sequence[SourceLineInfo],
+    blame_ranges: Sequence[GitHubFileBlameRange],
+    path: str,
+    repo_name: str,
+    ref: str,
+):
+    matching_file = [
+        f for f in files if f.path == path and f.repo.name == repo_name and f.ref == ref
+    ][0]
```

Comment:
Is it possible this could cause an index error if nothing in the loop matches the conditionals?

Reply:
Technically it shouldn't be possible, since the values here (path, repo_name, ref) come from the file mapping which was generated from the initial `files` list. If that ever isn't the case it means something has gone very wrong so I think failing fast it probably correct